Xplat/src/main/java/vazkii/botania/common/item/WorldshaperssSextantItem.java

```diff
@@ -112,32 +165,67 @@ public void releaseUsing(ItemStack stack, Level world, LivingEntity living, int
 			int x = ItemNBTHelper.getInt(stack, TAG_SOURCE_X, 0);
 			int y = ItemNBTHelper.getInt(stack, TAG_SOURCE_Y, Integer.MIN_VALUE);
 			int z = ItemNBTHelper.getInt(stack, TAG_SOURCE_Z, 0);
-			int iradius = (int) radius + 1;
+			boolean sphere = getSphereMode(stack);
 			if (y != Integer.MIN_VALUE) {
 				Map<BlockPos, IStateMatcher> map = new HashMap<>();
-				for (int i = 0; i < iradius * 2 + 1; i++) {
-					for (int j = 0; j < iradius * 2 + 1; j++) {
-						int xp = x + i - iradius;
-						int zp = z + j - iradius;
-
-						if ((int) Math.floor(MathHelper.pointDistancePlane(xp, zp, x, z)) == iradius - 1) {
-							map.put(new BlockPos(xp - x, 0, zp - z), matcher);
-						}
-					}
+				if (sphere) {
+					makeSphere(matcher, radius + 0.5, map);
+				} else {
+					makeCircle(matcher, radius + 0.5, map);
 				}
 				IMultiblock sparse = PatchouliAPI.get().makeSparseMultiblock(map).setId(MULTIBLOCK_ID);
-				Proxy.INSTANCE.showMultiblock(sparse, Component.literal("r = " + (int) radius), new BlockPos(x, y, z), Rotation.NONE);
+				Proxy.INSTANCE.showMultiblock(sparse, Component.literal("r = " + getRadiusString(radius)),
+						new BlockPos(x, y, z), Rotation.NONE);
 			}
 		}
 	}
 
-	private void reset(Level world, ItemStack stack) {
-		ItemNBTHelper.setInt(stack, TAG_SOURCE_Y, Integer.MIN_VALUE);
+	private static void makeCircle(IStateMatcher matcher, double radius, Map<BlockPos, IStateMatcher> map) {
+		// 2D version of makeSphere, assuming y=0 at all times
+		final int maxR2 = (int) Math.floor(radius * radius);
+		int z = (int) Math.floor(radius);
+		for (int x = 0;; x++) {
+			while (x * x + z * z > maxR2 && z >= x) {
+				z--;
+			}
+			if (z < x) {
+				break;
+			}
+			generateMirroredPositions(x, z, map, matcher);
+			generateMirroredPositions(z, x, map, matcher);
+		}
+	}
+
+	private static void generateMirroredPositions(int x, int z, Map<BlockPos, IStateMatcher> map, IStateMatcher matcher) {
+		Stream.of(
+				new BlockPos(x, 0, z), new BlockPos(-x, 0, z),
+				new BlockPos(x, 0, -z), new BlockPos(-x, 0, -z)
+		).forEach(pos -> map.put(pos, matcher));
+	}
+
+	private static boolean getSphereMode(ItemStack stack) {
+		return ItemNBTHelper.getBoolean(stack, TAG_SPHERE_MODE, false);
+	}
+
+	private void reset(Level world, Player player, ItemStack stack) {
+		if (ItemNBTHelper.getInt(stack, TAG_SOURCE_Y, Integer.MIN_VALUE) == Integer.MIN_VALUE) {
+			if (!world.isClientSide) {
+				setSphereMode(stack, !getSphereMode(stack));
+			} else {
+				player.playSound(BotaniaSounds.ding, 0.1F, 1F);
+			}
+		} else {
+			ItemNBTHelper.setInt(stack, TAG_SOURCE_Y, Integer.MIN_VALUE);
+		}
 		if (world.isClientSide) {
 			Proxy.INSTANCE.clearSextantMultiblock();
 		}
 	}
 
+	private static void setSphereMode(ItemStack stack, boolean sphereMode) {
+		ItemNBTHelper.setBoolean(stack, TAG_SPHERE_MODE, sphereMode);
```

Comment:
let's just store this as a string in `TAG_MODE` in case we want to add more shapes in the future